kingpin/actors/packagecloud.py

```diff
@@ -0,0 +1,358 @@
+                'http_methods': {'get': {}}
+            },
+            'delete': {
+                'path': 'repos/%account%/%repo%/%distro_version%/%filename%',
+                'http_methods': {'delete': {}}
+            },
+        },
+        'auth': {
+            'user': TOKEN,
+            'pass': ''
+        }
+    }
+
+
+class PackagecloudBase(base.HTTPBaseActor):
+
+    """Simple packagecloud Abstract Base Object"""
+
+    def __init__(self, *args, **kwargs):
+        """Check required environment variables."""
+        super(PackagecloudBase, self).__init__(*args, **kwargs)
+
+        if not ACCOUNT:
+            raise exceptions.InvalidCredentials(
+                'Missing the "PACKAGECLOUD_ACCOUNT" environment variable.')
+
+        if not TOKEN:
+            raise exceptions.InvalidCredentials(
+                'Missing the "PACKAGECLOUD_TOKEN" environment variable.')
+
+        rest_client = api.RestClient()
+        self._packagecloud_client = PackagecloudAPI(client=rest_client)
+
+    @gen.coroutine
+    def _get_all_packages(self, repo):
+        """Simple method for fetching a dictionary of all packages in a repo"""
+        all_packages = yield self._packagecloud_client.packages(
+            token=TOKEN, account=ACCOUNT, repo=repo).http_get()
+        raise gen.Return(all_packages)
+
+    @gen.coroutine
+    def _delete(self, packages_to_delete, repo, older_than=0,
+                number_to_keep=0):
+        """Generic packagecloud delete method, optionally supporting deleting
+        old packages by date."""
+
+        deleted_packages = []
+        all_packages = yield self._get_all_packages(repo=repo)
+
+        # Extract a list of unique package names
+        package_names = [package['name'] for package in all_packages]
+        package_names = set(package_names)
+
+        pattern = re.compile(packages_to_delete)
+
+        # Loop through each unique package
+        for name in package_names:
+            if not pattern.match(name):
+                continue
+
+            # Build a sorted list of all versions of a package
+            packages = [{
+                'created_at': datetime.datetime.strptime(
+                    package['created_at'], '%Y-%m-%dT%H:%M:%S.%fZ'),
+                'distro_version': package['distro_version'],
+                'filename': package['package_html_url'].split('/')[-1],
+                'name': package['name']
+            } for package in all_packages if package['name'] == name]
+
+            packages.sort(key=lambda x: x.get('created_at'), reverse=True)
+
+            # Delete individual packages if they meet our criteria
+            number_in_repo = len(packages)
+            for package in packages:
+                log.debug('Inspecting package %s' % package)
+
+                if number_in_repo > number_to_keep:
+                    if not older_than:
+                        pass
+                    else:
```

Comment:
This nested if/else/if/else bit is nasty and hard to understand. This should be something much simpler to read:

```
# If we've hit our limit of packages to keep, bail!
if number_to_keep <= number_in_repo:
  return

# If the package is younger than our minimum deletable package age, bail!
actual_package_age = (datetime.datetime.now() - package['created_at'])
min_package_age = datetime.timedelta(seconds=older_than)
if package_age <= min_package_age:
  return
```